docs/v1/api.md

```diff
@@ -0,0 +1,239 @@
+
+# Motivations
+
+Two different demos were given at [10/3/2016’s sig-service-catalog meeting](https://www.youtube.com/watch?v=Kfar5Uw7CRg), both extremely promising. Each differed in quite a few ways, including how consumers used Kubernetes resources to communicate.
+
+This document details a standard API, including a schema for standard
+Kubernetes-native resources, to:
+
+- Formalize an API around the service broker concept as a first-class Kubernetes
+  (or Kubernetes-like) API for all to develop against.
+- Follow established patterns in the Kubernetes core API for globally-available
+  and namespaced resources
+- Facilitate the development of Kubernetes-like controllers in support of the
+  API, without undermining competition or choice
+- Ensure a smooth, consistent UX for users of the API and implementors of the
+  controllers
+
+# API Precedents in Kubernetes
+
+Broadly speaking, the system detailed herein can list services and allow consumers to provision, bind, unbind and deprovision them.
+
+We've drawn inspiration from the [Persistent Volumes](http://kubernetes.io/docs/user-guide/persistent-volumes/) API and the [Ingress Controllers](http://kubernetes.io/docs/user-guide/ingress/) system. The remainder of this section explains these inspirations in further detail.
+
+## Persistent Volumes
+
+First, this problem space is similar to that of persistent volumes. We model the data structures herein after those in persistent volumes primarily because we believe that consistency is a virtue. More importantly, the community benefits when similar problems have similar solutions.
+
+Specifically, we believe that the following existing types provide a good model for the mechanics of how the services offered by a broker can be used in Kubernetes:
+
+- `StorageClass`
+- `PersistentVolume`
+- `PersistentVolumeClaim`
+
+## Ingress Controllers
+
+Today, operators choose their own implementation of an ingress controller. Kubernetes gives the operator the ability to specify any ingress controller, and the complete freedom to implement it themselves.
+
+Similarly, permitting operators to implement or select their own service catalog controller would encourage freedom of choice and diversity. The analog to an ingress controller for the service catalog system is called a _service catalog controller_. Like ingress, these controllers run inside a standard pod. Anyone can write and deploy a service catalog controller, with no changes necessary to the Kubernetes install.
+
+Service catalog controllers vary, however, from ingress controllers, because a cluster can consist of many different controllers, each of which handles actions on a subset of the available service-related resources. For example, a cluster can run a service controller to handle AWS RDS databases, and a completely separate one to handle AWS Elastic Load Balancer instances.
```

Comment:
@arschles I'm not following. Service controllers shouldn't really be aware of the service itself should they? Why would it need to know if the service is mysql vs mongo? All of those differences should be hidden behind the service broker, and in particular the SB API.  If we can't do that then I think there's something wrong with the SB API because it means its not generic enough and implementation details have to be exposed.

Even in your example of AWS RDS vs AWS ELBs, the controller shouldn't do anything different between those two - in both cases it calls the SB provision and bind operations at the appropriate time.  If more needs to be done then we need to extend the SB API to cover those cases. The SB API is meant to push all of the service-specific logic away from the controllers.